rust/src/lib.rs

```diff
@@ -131,89 +132,118 @@ pub struct Allocation {
     pub amount: u32,
 }
 
-pub(crate) fn instrument_enter<T>(
-    caller: Caller<T>,
+pub(crate) fn instrument_enter(
     _input: &[Val],
     _output: &mut [Val],
     ctx: Arc<Mutex<InstrumentationContext>>,
+    function_names: &FunctionNames
 ) -> anyhow::Result<()> {
-    let bt = WasmBacktrace::capture(caller);
-
-    if let Some(frame) = bt.frames().first() {
-        if let Ok(mut cont) = ctx.lock() {
-            cont.enter(frame)?;
-        }
+    let func_id = _input[0].unwrap_i32() as u32;
+    let printname = function_names.get(&func_id);
+    if let Ok(mut cont) = ctx.lock() {
+        cont.enter(func_id, printname)?;
     }
-
     Ok(())
 }
 
-pub(crate) fn instrument_exit<T>(
-    caller: Caller<T>,
+pub(crate) fn instrument_exit(
     _input: &[Val],
     _output: &mut [Val],
     ctx: Arc<Mutex<InstrumentationContext>>,
 ) -> Result<()> {
-    let bt = WasmBacktrace::capture(caller);
-
-    if let Some(frame) = bt.frames().first() {
-        if let Ok(mut cont) = ctx.lock() {
-            cont.exit(frame)?;
-        }
+    let func_id = _input[0].unwrap_i32() as u32;
+    if let Ok(mut cont) = ctx.lock() {
+        cont.exit(func_id)?;
     }
-
     Ok(())
 }
 
-pub(crate) fn instrument_memory_grow<T>(
-    caller: Caller<T>,
+pub(crate) fn instrument_memory_grow(
     input: &[Val],
     _output: &mut [Val],
     ctx: Arc<Mutex<InstrumentationContext>>,
 ) -> Result<()> {
     let amount_in_pages = input[0].unwrap_i32(); // The number of pages requested by `memory.grow` instruction
-    let bt = WasmBacktrace::capture(caller);
-
-    // TODO: this should scream and die loudly if it can't actually get ahold of the frame
-    if let Some(frame) = bt.frames().last() {
-        if let Ok(mut cont) = ctx.lock() {
-            cont.allocate(frame, amount_in_pages as u32)?;
-        }
+    if let Ok(mut cont) = ctx.lock() {
+        cont.allocate(amount_in_pages as u32)?;
     }
     Ok(())
 }
 
 const MODULE_NAME: &str = "dylibso_observe";
 
 type EventChannel = (Sender<Event>, Receiver<Event>);
+type FunctionNames = HashMap<u32, String>;
 
 /// Link observability import functions required by instrumented wasm code
-pub fn add_to_linker<T: 'static>(id: usize, linker: &mut Linker<T>) -> Result<EventChannel> {
+pub fn add_to_linker<T: 'static>(id: usize, linker: &mut Linker<T>, data : &Vec<u8>) -> Result<EventChannel> {
     let (ctx, events_tx, events_rx) = InstrumentationContext::new(id);
 
-    let t = FuncType::new([], []);
+    // Build up a table of function id to name mapping
+    //let mut function_names = Vec::new();
+    let mut function_names = FunctionNames::new();
+    let parser = wasmparser::Parser::new(0);
+    for payload in parser.parse_all(&data) {
+        let payload = payload?;
```

Comment:
Minor rust tip but not important, it's conventional to unwrap the results as you get them:

```rust
for payload in parser.parse_all(&data)? {
```

Reply:
```
error[E0277]: the `?` operator can only be applied to values that implement `std::ops::Try`
   --> rust/src/lib.rs:185:20
    |
185 |     for payload in parser.parse_all(&data)? {
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^ the `?` operator cannot be applied to type `impl Iterator<Item = std::result::Result<Payload<'_>, BinaryReaderError>>`
    |
    = help: the trait `std::ops::Try` is not implemented for `impl Iterator<Item = std::result::Result<Payload<'_>, BinaryReaderError>>`
```
Any suggestions instead?

Reply:
Oops yeah that's not going to work. I read that incorrectly You could put it on the match `match payload? {` but if that doesn't compile what you have works fine!